matrix_client/_api_asyncio.py

```diff
@@ -0,0 +1,78 @@
+import json
+from asyncio import sleep
+
+from matrix_client.api import MatrixHttpApi
+from matrix_client.errors import MatrixError, MatrixRequestError
+
+
+class AsyncHTTPAPI(MatrixHttpApi):
+    """
+    Contains all raw matrix HTTP client-server API calls using asyncio and aiohttp.
+
+    Usage:
+        async def main():
+            async with aiohttp.ClientSession() as session:
+                mapi = AsyncHTTPAPI("http://matrix.org", session)
+                resp = await mapi.get_room_id("#matrix:matrix.org")
+                print(resp)
+
+
+        loop = asyncio.get_event_loop()
+        loop.run_until_complete(main())
+    """
+
+    def __init__(self, base_url, client_session, token=None):
+        self.base_url = base_url
+        self.token = token
+        self.txn_id = 0
+        self.validate_cert = True
+        self.client_session = client_session
+
+    async def _send(self,
+                    method,
+                    path,
+                    content=None,
+                    query_params={},
+                    headers={},
+                    api_path="/_matrix/client/r0"):
+        if not content:
+            content = {}
+
+        method = method.upper()
+        if method not in ["GET", "PUT", "DELETE", "POST"]:
+            raise MatrixError("Unsupported HTTP method: %s" % method)
+
+        if "Content-Type" not in headers:
+            headers["Content-Type"] = "application/json"
+
+        if self.token:
+            query_params["access_token"] = self.token
+        endpoint = self.base_url + api_path + path
+
+        if headers["Content-Type"] == "application/json":
+            content = json.dumps(content)
+
+        while True:
+            request = self.client_session.request(
+                method,
+                endpoint,
+                params=query_params,
+                data=content,
+                headers=headers)
+            async with request as response:
+                if response.status < 200 or response.status >= 300:
+                    raise MatrixRequestError(
+                        code=response.status, content=await response.text())
+
+                if response.status == 429:
```

Comment:
This if statement is wrong, this should be first otherwise it's eaten by the previous one. See https://github.com/opsdroid/connector-matrix/commit/13d72309dfc27201bd55c808032b5065295f67c9

Reply:
So you gonna fix that up? How disappointing, that I finally find 20 minutes to
review this PR, and it isn't even ready. ;)